### ISSUE TITLE
luarocks.loader doesn't respect dependency contraints

### ISSUE THREAD

**Issue by lhemkendreis** [luarocks/luarocks#1175]
Trying to understand how luarocks works, I created an example with two rocks, rock `A` and rock `B`, where rock `B` exists in two versions `1.0` and `0.1` and rock `A`'s dependency is `B < 1.0`. There are no other dependencies specified. 

When I `require("A")` - which will then `require("B")` in its turn - I would expect the `0.1` of rock `B` to be loaded, since it is the only one which satisfies the dependency `B < 1.0`.

To my surprise, I can only ever load version `1.0` of rock `B`. Even more odd: When I inspected the content of the loader's "context" table, it associated rock `B` with the expected version `0.1`. So the loader had correctly determined that it had to load the version `0.1` but in the end it still loaded the highest version of rock `B` it could find.

I looked in the source code of `luarocks.loader` (on my system at _/usr/local/Cellar/luarocks/3.3.1/share/lua/5.3/luarocks/loader.lua_) and I found this bit:

```lua
if loader.context[name] == version then
  return name, version, file_name
end
```

This seems to be where the context table is queried. If the current values for `name` and `version` match an entry in the context table, then the relevant values are returned. So far, this makes sense to me.

The above lines are from the function `add_providers`, so lets see where that function is called:

```lua
   -- ... more stuff above
   local providers = {}
   local initmodule
   for _, tree in ipairs(tree_manifests) do
      local entries = tree.manifest.modules[module]
      if entries then
         add_providers(providers, entries, tree, module, filter_file_name)
      else
         initmodule = initmodule or module .. ".init"
         entries = tree.manifest.modules[initmodule]
         if entries then
            -- this next line is what was called during my tests:
            add_providers(providers, entries, tree, initmodule, filter_file_name)
         end
      end
   end

   if next(providers) then
      table.sort(providers, sort_versions)
      local first = providers[1]
      return first.name, first.version.string, first.module_name, first.tree
   end
```

As you can see, all return values from `add_providers` are discarded. This did not make a lot of sense to me, so I changed the code to to the following:

```lua
-- from this:
add_providers(providers, entries, tree, initmodule, filter_file_name)
-- to this:
local n, v, f = add_providers(providers, entries, tree, initmodule, filter_file_name)
if n then
  return n, v, f
end
```

And after that it worked as expected. Is this an error or what am I not getting here? :-( 

Thanks a lot :-) 

L.

**Comment by hishamhm** [comment]
Yes, this was a bug introduced in 65c417e0ecda55f44c691df032163a8c08f0b52a ! Great catch, thank you for the detailed description!

**Comment by hishamhm** [comment]
@lhemkendreis I pushed your suggested fix and added a regression test in #1180. Will wait  for the CI results then merge! Thank you again!

**Comment by lhemkendreis** [comment]
Awesome, thank you for the kind and quick response! You made this a very pleasant experience for me. :-)
